lib/HttpClient.ts

```diff
@@ -259,96 +273,136 @@ export class HttpClient {
         var defaultPort: number = usingSsl ? 443 : 80;
+        info.options = <http.RequestOptions>{};
+        info.options.host = info.parsedUrl.hostname;
+        info.options.port = info.parsedUrl.port ? parseInt(info.parsedUrl.port) : defaultPort;
+        info.options.path = (info.parsedUrl.pathname || '') + (info.parsedUrl.search || '');
+        info.options.method = method;
+        info.options.headers = headers || {};
+        info.options.headers["User-Agent"] = this.userAgent;
+        info.options.agent = this._getAgent(requestUrl);
 
-        let proxyConfig: ifm.IProxyConfiguration = this._httpProxy;
+        // gives handlers an opportunity to participate
+        if (this.handlers) {
+            this.handlers.forEach((handler) => {
+                handler.prepareRequest(info.options);
+            });
+        }
 
-        // fallback to http_proxy and https_proxy env
-        let https_proxy: string = process.env[EnvironmentVariables.HTTPS_PROXY];
-        let http_proxy: string = process.env[EnvironmentVariables.HTTP_PROXY];
+        return info;
+    }
 
-        if (!proxyConfig) {
-            if (https_proxy && usingSsl) {
-                proxyConfig = {
-                    proxyUrl: https_proxy
-                };
-            } else if (http_proxy) {
-                proxyConfig = {
-                    proxyUrl: http_proxy
-                };
-            }
+    private _getAgent(requestUrl: string) {
+        let agent;
+        var proxy = this._getProxy(requestUrl);
+        let useProxy = proxy.proxyUrl && proxy.proxyUrl.hostname && !this._isBypassProxy(requestUrl);
+
+        if (this._reuseAgent && useProxy) {
+            agent = this._proxyAgent;
         }
 
-        let proxyUrl: url.Url;
-        let proxyAuth: string;
-        if (proxyConfig) {
-            if (proxyConfig.proxyUrl.length > 0) {
-                proxyUrl = url.parse(proxyConfig.proxyUrl);
-            }
+        if (this._reuseAgent && !useProxy) {
+            agent = this._agent;
+        }
 
-            if (proxyConfig.proxyUsername || proxyConfig.proxyPassword) {
-                proxyAuth = proxyConfig.proxyUsername + ":" + encodeURIComponent(proxyConfig.proxyPassword);
-            }
+        // if agent is already assigned use that agent.
+        if (!!agent) {
+            return agent;
         }
 
-        info.options = <http.RequestOptions>{};
-        info.options.host = info.parsedUrl.hostname;
-        info.options.port = info.parsedUrl.port ? parseInt(info.parsedUrl.port) : defaultPort;
-        info.options.path = (info.parsedUrl.pathname || '') + (info.parsedUrl.search || '');
-        info.options.method = method;
-        info.options.headers = headers || {};
-        info.options.headers["User-Agent"] = this.userAgent;
+        var parsedUrl = url.parse(requestUrl);
+        let usingSsl = parsedUrl.protocol === 'https:';
+        let keepAlive = false;
+        let maxSockets = http.globalAgent.maxSockets;
```

Comment:
Let's get rid of the http.globalAgent.maxSockets since you added the option.  Use the option

Reply:
http.globalAgent.maxSockets is the default value, if value in options is specified it will be overridden below